### PR TITLE
Fix issues running initial setup in dev mode

### PR DESCRIPTION
**Pull request by kzantow**
Minor tweaks to make it easier to run & work on the setup wizard when running in development mode:
- set the Jenkins version based on the maven pom
- add option to set initial password from the command line

@jenkinsci/code-reviewers

### CODE REVIEW

**Comment by daniel-beck** [comment]
@jenkinsci/code-reviewers Any opinions on whether this is a good idea? Maybe we should only consider the system property with a check whether we're running in development mode? Or could this be a legitimate user feature e.g. for mass deployments/Jenkins as a service?

**Comment by oleg-nenashev** [comment]
I think it makes sense to start with Mock Security realm and admin:admin if Jenkins runs in the dev mode.

My gut-feeling is that DEV-mode is useful, but the current check implementation may cause security risks once the usage of the mode gets extended. Jenkins does not control "JENKINS_HOME/..." at all, so on a production instance a malicious attacker may potentially have an access to it. I would propose to just use a system property, which activates the development mode. Then you can alter the security setup logic for it.

@kzantow @daniel-beck WDYT?

**Comment by kzantow** [comment]
@oleg-nenashev I'm confused... if an attacker has access to JENKINS_HOME, they can do all kinds of bad things, irrespective of the changes here. Specifically, this will lookup a version number from ../pom.xml IFF "${build.version}" is the version number and allow the user to set an initial installer password when running Jenkins. Also, you can add a command line parameter to specify the initial install password instead of a generated password, which is only used if the initial installer runs. For the most part, if the user has access to JENKINS_HOME, and, more specifically, to the running Jenkins process, I assume they'd be able to get the generated password anyway, right?

**Comment by daniel-beck** [comment]
@oleg-nenashev I don't understand how your comment relates to what this PR does. It allows defining a default installer password, and sets the Jenkins version in dev mode. That's it.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -4608,6 +4609,20 @@ private static void computeVersion(ServletContext context) {
         }
         String ver = props.getProperty("version");
         if(ver==null)   ver = UNCOMPUTED_VERSION;
+        if("${build.version}".equals(ver)) {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: :bug: use

```java
if (Main.isDevelopmentMode) {
```

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 2]
```diff
@@ -4608,6 +4609,20 @@ private static void computeVersion(ServletContext context) {
...
+            // in dev mode?
+            try {
+                File pom = new File("../pom.xml");
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: :bug: inappropriate assumption about CWD of `mvn` process.

**Comment by jglick** [comment]
> set the Jenkins version based on the maven pom

I do not even understand what problem you are trying to solve here. In `master` I apply the following patch

```diff
diff --git a/core/src/main/java/jenkins/model/Jenkins.java b/core/src/main/java/jenkins/model/Jenkins.java
index 9e5bdf8..e834e16 100644
--- a/core/src/main/java/jenkins/model/Jenkins.java
+++ b/core/src/main/java/jenkins/model/Jenkins.java
@@ -4540,6 +4540,7 @@ public class Jenkins extends AbstractCIBase implements DirectlyModifiableTopLeve
         }
         String ver = props.getProperty("version");
         if(ver==null)   ver="?";
+        System.err.println("TODO running version " + ver);
         VERSION = ver;
         context.setAttribute("version",ver);
 
```

and then

    mvn -DskipTests clean install && mvn -f war hudson-dev:run

and I see as expected

    TODO running version 1.655-SNAPSHOT (private-03/23/2016 13:35 GMT-jglick)

If you are seeing `${build.version}` then you are doing something wrong, since [this code](https://github.com/jenkinsci/jenkins/blob/5999996aa36ea1429dc5668f18598384f3405aba/pom.xml#L273-L303) already processes [this definition](https://github.com/jenkinsci/jenkins/blob/5999996aa36ea1429dc5668f18598384f3405aba/core/src/filter/resources/jenkins/model/jenkins-version.properties#L1), but I would need to know how to reproduce in order to advise further. Perhaps just incorrect IDE configuration? Works fine from the command line or [from NetBeans](https://github.com/jenkinsci/jenkins/blob/5999996aa36ea1429dc5668f18598384f3405aba/war/nbactions.xml#L30-L35).

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 2 above)
Comment (kzantow) [inline core/src/main/java/jenkins/model/Jenkins.java]: This is run from the war project, cwd is known.

**Comment by kzantow** [comment]
@jglick it does not work fine when run inside Eclipse when using jenkins-dev:run, I suppose that's the main case I'd like to solve. I could fix Eclipse to better support the build-helper plugin, but seeing as this change should cause no problems to anyone as it's written, and I know nothing about Eclipse development, this seems like an easy enough fix. Maybe you have a better idea?

**Comment by jglick** [comment]
> could this be a legitimate user feature e.g. for mass deployments/Jenkins as a service?

`System.getProperty("jenkins.install.initialAdminPassword", …)` does not seem sufficiently secure to use in a production setting, even assuming you had some deployment tool that generated a randomized password, and the Jenkins instance was otherwise properly secured (for example, no master executors): it is too easy to leak system properties. Yes `/systemInfo` is guarded by `Jenkins.ADMINISTER`, but I fear that various plugins (EnvInject…) might inadvertently expose them to malicious jobs or CLI/API requests.

If you are setting up a scripted deployment, I think it would be much wiser to use the `system-config-dsl` plugin (or one of its many predecessors based on XML file creation, or a plain old `Dockerfile`) to define a regular security realm that uses a more secure means of storing its password; `HudsonPrivateSecurityRealm.Details` in fact does not store the password at all, only a jBCrypt hash, so even an inadvertent leak of `$JENKINS_HOME/users/admin/config.xml` (plus probably some other salt files) would at worst allow dictionary attacks.

**Comment by jglick** [comment]
> I could fix Eclipse to better support the build-helper plugin

Not being an Eclipse user, I can only suggest looking at [code like this](https://github.com/jenkinsci/jenkins/blob/5999996aa36ea1429dc5668f18598384f3405aba/pom.xml#L863-L893).

**Comment by jtnord** [comment]
eclipse puts the `src/main/resources` on the classpath - so you get the unfiltered resources (if filtering is enabled).  Yeah eclipse support for Maven is sucky but they have finally landed (well a PR) for multiple classpath types so that test scope code from one project will not leak to production code in another...
manual workaround when launching the test is to fudge the classpath and add target/classes/

**Comment by jglick** [comment]
> eclipse puts the `src/main/resources` on the classpath

FWIW NetBeans uses `target/classes/` but then synchronizes _changes_ from `src/main/java/` or `src/main/resources/`, which works well at least for this kind of case.

**Pull request closed without merging**